File: SamAndSubstring/Python/SamAndSubstring.py
#!/bin/python3

MOD = 1000000007

if __name__ == '__main__':
    n = input()
    result = 0
    frequency = 1

    for i in range(len(n), 0, -1):
        result = (result + (int(n[i - 1])) * frequency * i) % MOD
        frequency = (frequency * 10 + 1) % MOD

    print(str(result))

File: SamAndSubstring/CPP/SamAndSubstring.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

#define mod 1000000007;

int substrings(string n)
{
    long long result = 0;
    long long frequency = 1;

    for (int i = n.size() - 1; i >= 0; i--)
    {
        result = (result + (n[i] - '0') * frequency * (i + 1)) % mod;
        frequency = (frequency * 10 + 1) % mod;
    }

    return result;
}

int main()
{
    string n;
    cin >> n;

    int result = substrings(n);

    cout << result << "\n";

    return 0;
}